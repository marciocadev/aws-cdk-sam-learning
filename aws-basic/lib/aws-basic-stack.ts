import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Table, AttributeType } from '@aws-cdk/aws-dynamodb';
import { RestApi, LambdaIntegration } from '@aws-cdk/aws-apigateway';

export class AwsBasicStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new Table(this, 'BasicTable', {
      tableName: 'basic-db',
      partitionKey: {
        name: id,
        type: AttributeType.STRING
      }
    });

    const postLmb = new NodejsFunction(this, 'PostLambda', {
      functionName: 'basic-post',
      entry: path.join(__dirname, '/../src/post/index.ts')
    });
    
    table.grantWriteData(postLmb);

    const rest = new RestApi(this, 'BasicRest', {
      restApiName: 'basic-gateway'
    });
    const resource = rest.root.addResource('basic')
    resource.addMethod("post", new LambdaIntegration(postLmb));
  }
}
