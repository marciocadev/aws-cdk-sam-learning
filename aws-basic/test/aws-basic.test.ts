import { Capture, Match, Matcher, MatchResult, Template } from '@aws-cdk/assertions';
import { App } from '@aws-cdk/core';
import { AwsBasicStack } from '../lib/aws-basic-stack';

test('Snapshot', () => {
  // GIVEN
  const app = new App();
  // WHEN
  const stack = new AwsBasicStack(app, 'MyTestStack');
  // THEN
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});

describe('Lambda Nodejs tests', () => {
  let app:any;
  let stack:any;
  let template:Template;

  beforeAll(() => {
    process.env = {
      tableName: 'basic-db',
      AWS_SAM_LOCAL: 'true'
    };
    // GIVEN
    app = new App();
    // WHEN
    stack = new AwsBasicStack(app, 'MyTestStack');
    // THEN
    template = Template.fromStack(stack);
  });

  test('Lambda Exists', () => {
    template.findResources('AWS::Lambda::Function');
  });

  test('DynamoDB Policy Permition', () => {

    template.findResources('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: [{
            Action: ["dynamodb:BatchWriteItem", "dynamodb:PutItem",
              "dynamodb:UpdateItem", "dynamodb:DeleteItem"],
            Effect: 'Allow'
          }
        ]
      }
    });

    // template.hasResourceProperties('AWS::IAM::Policy', {
    //   PolicyDocument: {
    //     Statement: Match.arrayWith([{
    //       Action: ["dynamodb:BatchWriteItem",
    //       "dynamodb:PutItem",
    //       "dynamodb:UpdateItem",
    //       "dynamodb:DeleteItem",]
    //     }])
    //   }
    // })
  });
  
});
