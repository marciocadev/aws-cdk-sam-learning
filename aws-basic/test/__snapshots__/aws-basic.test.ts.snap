// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aArtifactHash46D722E7": Object {
      "Description": "Artifact hash for asset \\"837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87a\\"",
      "Type": "String",
    },
    "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aS3Bucket93DEF033": Object {
      "Description": "S3 bucket for asset \\"837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87a\\"",
      "Type": "String",
    },
    "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aS3VersionKey32CF0C18": Object {
      "Description": "S3 key for asset version \\"837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87a\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BasicTableD462653D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "MyTestStack",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "MyTestStack",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "basic-db",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "PostLambda0EF1C7F9": Object {
      "DependsOn": Array [
        "PostLambdaServiceRoleDefaultPolicy1E8FF5CA",
        "PostLambdaServiceRole1CD56985",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aS3Bucket93DEF033",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aS3VersionKey32CF0C18",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters837adb1ed2a9a0e09cd6a966621962463a7221c732709840571f0cd76e75a87aS3VersionKey32CF0C18",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "basic-post",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PostLambdaServiceRole1CD56985",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostLambdaServiceRole1CD56985": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostLambdaServiceRoleDefaultPolicy1E8FF5CA": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "BasicTableD462653D",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostLambdaServiceRoleDefaultPolicy1E8FF5CA",
        "Roles": Array [
          Object {
            "Ref": "PostLambdaServiceRole1CD56985",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
